### PR TITLE
 Hack to fix non-bmp unicode input for windows

### LINKED ISSUE #1477: Exception on non-BMP character input in Windows Terminal
When attempting to input a non-BMP character, like 😀, in the (new) Windows Terminal, the following exception is thrown and the input fails:

```
Unhandled exception in event loop:
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\input\win32.py", line 512, in ready
    callback()
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\application\application.py", line 662, in read_from_input
    keys = self.input.read_keys()
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\input\win32.py", line 75, in read_keys
    return list(self.console_input_reader.read())
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\input\win32.py", line 233, in read
    all_keys = list(self._get_keys(read, input_records))
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\input\win32.py", line 292, in _get_keys
    for key_press in self._event_to_key_presses(ev):
  File "c:\users\ryan\appdata\local\programs\python\python39\lib\site-packages\prompt_toolkit\input\win32.py", line 330, in _event_to_key_presses
    ascii_char = u_char.encode("utf-8")

Exception 'utf-8' codec can't encode character '\ud83d' in position 0: surrogates not allowed
```
It seems that the terminal is sending the character as separate utf-16 surrogate codes, which breaks the win32 input logic. Either the two key events would somehow have to be merged into one KeyPress with the full character or the unicode would have to be normalized somehow later on.

### PR DESCRIPTION
**Pull request by rmccampbell**
Fixes #1477. This transforms the list of decoded KeyPress events and merges surrogates into single Unicode characters.

### CODE REVIEW

File: prompt_toolkit/input/win32.py [hunk 1]
```diff
@@ -315,6 +318,37 @@ def _get_keys(
                     for key_press in self._handle_mouse(ev):
                         yield key_press
 
+    @staticmethod
+    def _merge_paired_surrogates(key_presses: List[KeyPress]) -> Iterator[KeyPress]:
```
Comment (jonathanslenders) [inline prompt_toolkit/input/win32.py:322]: Just wondering, is this similar to what ``unicodedata.normalize`` does, but for ``KeyPress`` objects?

File: prompt_toolkit/input/win32.py [hunk 2]
```diff
@@ -260,7 +263,7 @@ def read(self) -> Iterable[KeyPress]:
                 # Pasting: if the current key consists of text or \n, turn it
                 # into a BracketedPaste.
                 data = []
-                while k and (isinstance(k.key, str) or k.key == Keys.ControlJ):
+                while k and (not isinstance(k.key, Keys) or k.key == Keys.ControlJ):
```
Comment (jonathanslenders) [inline prompt_toolkit/input/win32.py:266]: Is this change needed?

Comment (rmccampbell) [reply prompt_toolkit/input/win32.py:266]: This is unrelated to the main change but a potential bug fix. Because the `Keys` enum inherits `str`, technically the original check would always pass, so I think this change more correctly matches the intent of the code. Actually it would probably be better to make the enum not inherit `str` in the first place (which makes the enum a drop-in replacement for str in most cases); I don't think this is very useful but it may be relied on by some code so not sure if it would be safe to change without analyzing more of the code base.

File: prompt_toolkit/input/win32.py (hunk 1 above)
Comment (rmccampbell) [reply prompt_toolkit/input/win32.py:322]: Not exactly, although a similar idea. unicodedata.normalize works at a higher level of abstraction, between "conceptual" characters and code points, i.e. it can merge 'a' and '◌̈' into 'ä'. This function is merging UTF-16 surrogate pair code units (see https://en.wikipedia.org/wiki/UTF-16) into single code points. Apparently the underlying windows API calls are working in UTF-16 and treating the code units as individual characters, when it ideally should return full code points, hence this workaround.

Comment (rmccampbell) [reply prompt_toolkit/input/win32.py:322]: Example: the character 😀 (U+1F600) can be encoded with the two 16-bit code units [0xD83D, 0xDE00].

**Comment by jonathanslenders** [comment]
@rmccampbell: Thank you for this!
Merged from here: https://github.com/prompt-toolkit/python-prompt-toolkit/pull/1482 (squashed/reformatted).
